### PR TITLE
Use vertical formatting when function args width > limit

### LINKED ISSUE #342: Some kind of max width heuristic for function calls
I find very long function calls on one line hard to read and ugly. I wonder if we should have some max with heuristic which forces them into vertical formatting before we hit the hard width limit? This seems to be the source of a large percentage of the 'poor decisions' I see rustfmt making when running on rustc crates.

### PR DESCRIPTION
**Pull request by nrc**
Closes #342

### CODE REVIEW

File: src/lists.rs [hunk 1]
```diff
@@ -26,6 +26,8 @@ pub enum ListTactic {
     Horizontal,
     // Try Horizontal layout, if that fails then vertical
     HorizontalVertical,
+    // HorizontalVertical with a soft limit.
```
Comment (marcusklaas) [inline src/lists.rs]: Maybe make explicit that this is a character limit. An item limit would also be feasible.

File: src/types.rs [hunk 2]
```diff
@@ -363,7 +365,8 @@ impl Rewrite for ast::TyParamBound {
                 tref.rewrite(context, width, offset)
             }
             ast::TyParamBound::TraitTyParamBound(ref tref, ast::TraitBoundModifier::Maybe) => {
-                Some(format!("?{}", try_opt!(tref.rewrite(context, width - 1, offset + 1))))
+                Some(format!("?{}",
+                             try_opt!(tref.rewrite(context, width - 1, offset + 1))))
```
Comment (marcusklaas) [inline src/types.rs]: Not really relevant for this PR, but we should probably have a checked subtraction here at some point.

**Comment by marcusklaas** [comment]
Looks good! Is this something we should consider for struct literals as well?

Otherwise, r+.

**Comment by nrc** [comment]
Yeah, we should do it for struct lits too, but I thought I'd leave it for another PR (I want to think about number of items there too).

**Pull request merged**